File: 94. Binary Tree Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        res = []
        stack = [root]
        while stack:
            while root and root.left:
                stack += [root.left]
                root = root.left
            root = stack.pop()
            res += [root.val]
            if root.right:
                stack += [root.right]
            root = root.right
        return res

File: 94. Binary Tree Inorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        if (root == NULL)
            return {};
        stack <TreeNode*> stack;
        stack.push(root);
        vector <int> res;
        while (!stack.empty())
        {
            while (root != NULL && root->left != NULL)
            {
                stack.push(root->left);
                root = root->left;
            }
            root = stack.top();
            stack.pop();
            res.push_back(root->val);
            if (root->right)
                stack.push(root->right);
            root = root->right;
        }
        return res;
    }
};
